(DOCSP-11147): Initted Realm Electron Quickstart Doc

mongomoe opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-11147

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/realm/mohammad.hunan/DOCSP-11147-Realm-Electron-QuickStart/node/electron.html

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+   run();   
+
+
+On line 27, a non-synced realm is opened in this rendere r process. All
```
```suggestion
On line 27, a non-synced realm is opened in this renderer process. All
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+
+On line 27, a non-synced realm is opened in this rendere r process. All
+write transactions on this local realm, will not be synced by default. In order to sync these writes, the `ipcRenderer <https://www.electronjs.org/docs/api/ipc-renderer>` Electron API is used to emit an "asynchronous-message" with the arguement "sync" on line 40. This tells our ``main process`` to upload all local changes to the synced realm. 
+
```
This page could use a summary.

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+title: Install CRACO to Alter Your Webpack Configuration
+ref: install-craco
+content: |
+  To allow your application to work properly with electron you have to
```
```suggestion
  To allow your application to work properly with Electron, you have to
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  To allow your application to work properly with electron you have to
+  alter your webpack configuration. By default, applications created via
+  create-react-app use a webpack file that is preconfigured and hidden
+  to the end user. Using CRACO you can override these default
```
Link to CRACO?

mongomoe replied:
added 👍

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  To allow your application to work properly with electron you have to
+  alter your webpack configuration. By default, applications created via
+  create-react-app use a webpack file that is preconfigured and hidden
+  to the end user. Using CRACO you can override these default
```
>to the end user.

You could explain that unfortunately, this is not compatible with Realm and needs to be overridden.

mongomoe replied:
yup that makes sense, done 👍

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  specified to prevent all node_modules from being bundled. The
+  ``allowList`` key in the object passed  in to ``nodeExternals``
+  specifies a list of modules to include in the bundle. Here, webpack
+  and the electron-devtools and included in the bundle. In order to make
```
```suggestion
  and the electron-devtools are included in the bundle. In order to make
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+title: Install Electron
+ref: install-electron
+content: |
+  Install electron as a devDependency in order to run your application
```
```suggestion
  Install Electron as a devDependency in order to run your application
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+ref: install-electron
+content: |
+  Install electron as a devDependency in order to run your application
+  from a chromium desktop window using Electron. Run the following
```
This is a bit awkward. Surely people know what they're installing Electron for. Why not just say:

>To add Electron to your project, run the following command:

mongomoe replied:
that does sound better, fixed 👍

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+title: Create Your Electron Main Process File
+ref: create-your-electron-main-process
+content: |
+  The electron ``main process`` file can be thought of as the entry
```
```suggestion
  The Electron ``main process`` file can be thought of as the entry
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  electron.
+
+  .. note::
+     Each electron application can only have one ``main process``. The
```
```suggestion
     Each electron application can only have one **main process**. The
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+
+  .. note::
+     Each electron application can only have one ``main process``. The
+     main process can create web pages, and each web page runs in it's
```
```suggestion
     main process can create web pages. Each web page runs in its
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  .. note::
+     Each electron application can only have one ``main process``. The
+     main process can create web pages, and each web page runs in it's
+     own process, known as the ``renderer process``. To learn more about
```
```suggestion
     own process, known as a **renderer process**. To learn more about
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+     Each electron application can only have one ``main process``. The
+     main process can create web pages, and each web page runs in it's
+     own process, known as the ``renderer process``. To learn more about
+     this read the official `Electron Application
```
```suggestion
     this, read the official `Electron Application
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+  Add the following code to a new file called ``electron.js`` in the
+  ``public`` directory:
+
+  .. code-block:: javascript
```
Can you move this to a literalinclude?

mongomoe replied:
done 👍

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,175 @@
+
+     import Realm from "realm";
+
+  Use realm in the main process by adding the following to the top of the ``public/electron.js`` file (you will also need to require it in whichever file you write code using Realm in):
```
I would remove that parenthetical:
```suggestion
  Use Realm in the main process by adding the following to the top of the ``public/electron.js`` file:
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+
+
+On line 27, a non-synced realm is opened in this rendere r process. All
+write transactions on this local realm, will not be synced by default. In order to sync these writes, the `ipcRenderer <https://www.electronjs.org/docs/api/ipc-renderer>` Electron API is used to emit an "asynchronous-message" with the arguement "sync" on line 40. This tells our ``main process`` to upload all local changes to the synced realm. 
```
Broken link:
```suggestion
write transactions on this local realm, will not be synced by default. In order to sync these writes, the `ipcRenderer <https://www.electronjs.org/docs/api/ipc-renderer>`__ Electron API is used to emit an "asynchronous-message" with the arguement "sync" on line 40. This tells our ``main process`` to upload all local changes to the synced realm. 
```

mongomoe replied:
Thanks for pointing this out - fixed

cbush reviewed (changes requested):
Good work! I think the layout of the doc is pretty good. Research seems thorough and I think people will get good use out of it.

I went ahead and added a ton of suggestions to save you some time on some punctuation fixes. Please ask if you have any questions about my suggestions.

What needs to change:
- Please accept suggestions.
- Please move the code snippets into literalincludes. This is going to happen soon anyway so let's save ourselves the tech debt.
- Please move the setup headers into a subsection specifically for setup. I'm concerned the "Working Realm Using Renderer and Main Processes" section is getting lost otherwise.
- Please add a note or paragraph or sentence near the top that helps people decide which of the setup sections to read. You could also have a sentence in each setup section that lets people know whether they should read that section or if they can skip it. It's not clear from the section titles alone right now.
- Please write a summary section.

Thanks!

mongomoe commented:
@cbush per your top level comment ("Please accept suggestions.
Please move the code snippets into literalincludes. This is going to happen soon anyway so let's save ourselves the tech debt.
Please move the setup headers into a subsection specifically for setup. I'm concerned the "Working Realm Using Renderer and Main Processes" section is getting lost otherwise.
Please add a note or paragraph or sentence near the top that helps people decide which of the setup sections to read. You could also have a sentence in each setup section that lets people know whether they should read that section or if they can skip it. It's not clear from the section titles alone right now.
Please write a summary section."):

* Moved code snippets into literalincludes: ✅  (only moved the codesnippets that were full js files) 
* Moved the setup headers into subsections: ✅ 
* Added a paragraph at the top that helps people decide which setup to read & added a sentence in each set up subsection: ✅ 
* Added a summary:  ✅

mongomoe commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+applications using Electron and Realm:
+
+- 2 different processes cannot access a synced realm. Due to this, the recommended course of action is to open a non-synced realm on renderer processes, and open a synced realm on the main process. 
+- Since a non-synced realm is opened on the renderer process, the main process must be notified when a write transaction happens during the renderer process. 
```
Hmm, I prefer this current approach of:

- <challenge>, <solution> 

Where we present two challenges (2 diff processes can't access a synced realm, and a non-synced realm is opened on a renderer process). I think separating the two lists might dilute the fact that a synced realm can be opened on a renderer process (but is not recommended, due to the first bullet). What do. you think?

cbush replied:
This section is not clear as is. Please rework it.

mongomoe commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+
+.. include:: /includes/steps/realm-with-electron-using-create-react-app.rst
+
+Working Realm Using Renderer and Main Processes
```
Fixed

mongomoe commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+Setup Realm with Electron
+-------------------------
+
+.. include:: /includes/steps/realm-with-electron.rst
```
Added an intro sentence to these subsections

mongomoe commented on source/node/electron.txt:
```diff
@@ -0,0 +1,166 @@
+line 42, the `ipcMain <https://www.electronjs.org/docs/api/ipc-main>`_
+Electron API is used to listen for an "asynchronous-message" event. When
+that occurs, and the argument "sync" is passed in, all local changes
+will be uploaded to the synced realm. 
```
Good catch - fixed now

cbush commented on source/includes/steps-realm-with-electron.yaml:
```diff
@@ -0,0 +1,95 @@
+  In order to display HTML to the users, your application needs a
+  browser window. Create a `browser window
+  <https://www.electronjs.org/docs/api/browser-window>`_ using the
+  electron API. Copy the following code into your ``main.js`` file:
```
```suggestion
  Electron API. Copy the following code into your ``main.js`` file:
```

cbush commented on source/includes/steps-realm-with-electron-using-create-react-app.yaml:
```diff
@@ -0,0 +1,135 @@
+title: Run Your Application
+ref: run-your-app
+content: |
+  In order to run your application, specify a homepage and a main entry point for electron by adding the following to your ``package.json`` file: 
```
```suggestion
  In order to run your application, specify a homepage and a main entry point for Electron by adding the following to your ``package.json`` file: 
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+This presents the following unique challenges when developing
+applications using Electron and Realm:
+
+- Two different processes cannot access a synced realm. Due to this, the recommended course of action is to open a non-synced realm on renderer processes and open a synced realm on the main process.
```
This is still not as clear as it could be. At the very least:
```suggestion
- Two different processes cannot access a synced realm. Due to this, the recommended course of action is to only open non-synced realms on renderer processes and only one synced realm on the main process.
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+  :emphasize-lines: 39, 42, 45
+  :linenos:
+
+Notice that on line 38, the main process opens a synced realm. On line
```
Exactly what I warned about has happened: line 38 is blank:

![image](https://user-images.githubusercontent.com/20050130/95209220-202b0800-07b8-11eb-9ab0-7ad35e1f9014.png)

```suggestion
Notice that on line 39, the main process opens a synced realm. On line
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+asynchronous message occurs with the “sync” argument, the main process
+uploads all local changes.
+
+Consider the adjacent ``renderer process`` code:
```
Why this word?
```suggestion
Consider the ``renderer process`` code:
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+  :linenos:
+
+Notice that on line 38, the main process opens a synced realm. On line
+42, the `ipcMain <https://www.electronjs.org/docs/api/ipc-main>`_
```
```suggestion
45, the `ipcMain <https://www.electronjs.org/docs/api/ipc-main>`_
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+  :emphasize-lines: 28, 40
+  :linenos:
+
+On line 27, the renderer process opens a non-synced realm. All write
```
```suggestion
On line 28, the renderer process opens a non-synced realm. All write
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+  :linenos:
+
+On line 27, the renderer process opens a non-synced realm. All write
+transactions on this local Realm, will not be synced by default. In
```
```suggestion
transactions on this local realm will not be synced by default. In
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+transactions on this local Realm, will not be synced by default. In
+order to sync these writes, the `ipcRenderer
+<https://www.electronjs.org/docs/api/ipc-renderer>`_ Electron API emits
+an "asynchronous-message" with the argument "sync" on line 40. This
```
```suggestion
an "asynchronous-message" with the argument "sync" on line 42. This
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+-------
+
+- You can create an Electron Application using {+service+} by installing `Electron <https://www.npmjs.com/package/electron>`_ and the `MongoDB Realm JS SDK <https://www.npmjs.com/package/realm>`_.
+- You can create React Application with Realm using Electron by utilizing `craco <https://www.npmjs.com/package/@craco/craco>`_, Electron, and Realm.
```
```suggestion
- You can create a React application that uses Realm and Electron with `craco <https://www.npmjs.com/package/@craco/craco>`_.
```

cbush commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+- You can create an Electron Application using {+service+} by installing `Electron <https://www.npmjs.com/package/electron>`_ and the `MongoDB Realm JS SDK <https://www.npmjs.com/package/realm>`_.
+- You can create React Application with Realm using Electron by utilizing `craco <https://www.npmjs.com/package/@craco/craco>`_, Electron, and Realm.
+- Since 2 different processes cannot access a synced realm, open a non-synced realm on the renderer processes and a synced realm on the main process.
+- Notify the main process that a write transaction has occurred on the renderer process by utilizing the `ipcMain <https://www.electronjs.org/docs/api/ipc-main>`_ and the `ipcRenderer <https://www.electronjs.org/docs/api/ipc-renderer>`_ APIs.
```
Avoid overwriting:
```suggestion
- Notify the main process that a write transaction has occurred on the renderer process with the `ipcMain <https://www.electronjs.org/docs/api/ipc-main>`_ and the `ipcRenderer <https://www.electronjs.org/docs/api/ipc-renderer>`_ APIs.
```

cbush reviewed (approved):
Looks good. Please accept suggestions before merging. Thanks!

mongomoe commented on source/node/electron.txt:
```diff
@@ -0,0 +1,103 @@
+
+.. include:: /includes/steps/realm-with-electron-using-create-react-app.rst
+
+Working Realm Using Renderer and Main Processes
```
```suggestion
Working with Realm Using Renderer and Main Processes
```
